src/main/java/xades4j/providers/impl/HttpTimeStampTokenProvider.java

```diff
@@ -16,128 +16,35 @@
-
-    @Inject
-    public HttpTimeStampTokenProvider(MessageDigestEngineProvider messageDigestProvider, TSAHttpData tsaHttpData)
-    {
-        this.messageDigestProvider = messageDigestProvider;
-        this.tsaHttpData = tsaHttpData;
-        this.tsRequestGenerator = new TimeStampRequestGenerator();
-        this.tsRequestGenerator.setCertReq(true);
-
-        if (this.tsaHttpData.getUsername() != null)
-        {
-            String usrAndPwd = this.tsaHttpData.getUsername() + ":" + this.tsaHttpData.getPassword();
-            this.base64TsaUsrAndPwd = Base64.encodeBytes(usrAndPwd.getBytes());
-        }
-        else
-        {
-            this.base64TsaUsrAndPwd = null;
-        }
-    }
-
-    @Override
-    public final TimeStampTokenRes getTimeStampToken(
-            byte[] tsDigestInput,
-            String digestAlgUri) throws TimeStampTokenGenerationException
-    {
-        try
-        {
-            MessageDigest md = messageDigestProvider.getEngine(digestAlgUri);
-            byte[] digest = md.digest(tsDigestInput);
-
-            TimeStampRequest tsRequest = this.tsRequestGenerator.generate(
-                    identifierForDigest(digestAlgUri),
-                    digest,
-                    BigInteger.valueOf(System.currentTimeMillis()));
-            InputStream responseStream = getResponse(tsRequest.getEncoded());
-            TimeStampResponse tsResponse = new TimeStampResponse(responseStream);
-
-            if (tsResponse.getStatus() != PKIStatus.GRANTED
-                    && tsResponse.getStatus() != PKIStatus.GRANTED_WITH_MODS)
-            {
-                throw new TimeStampTokenGenerationException("Time stamp token not granted. " + tsResponse.getStatusString());
-            }
-            tsResponse.validate(tsRequest);
+public class HttpTimeStampTokenProvider extends AbstractTimeStampTokenProvider {
 
-            TimeStampToken tsToken = tsResponse.getTimeStampToken();
-            return new TimeStampTokenRes(tsToken.getEncoded(), tsToken.getTimeStampInfo().getGenTime());
-        } catch (UnsupportedAlgorithmException ex)
-        {
-            throw new TimeStampTokenGenerationException("Digest algorithm not supported", ex);
-        } catch (TSPException ex)
-        {
-            throw new TimeStampTokenGenerationException("Invalid time stamp response", ex);
-        } catch (IOException ex)
-        {
-            throw new TimeStampTokenGenerationException("Encoding error", ex);
-        }
+    public HttpTimeStampTokenProvider(MessageDigestEngineProvider messageDigestProvider, TSAHttpData tsaHttpData) {
+        super(messageDigestProvider, tsaHttpData);
     }
 
-    private InputStream getResponse(byte[] encodedRequest) throws TimeStampTokenGenerationException
-    {
-        try
-        {
-            HttpURLConnection connection = createHttpConnection();
+    protected byte[] getResponseFromServer(byte[] encodedRequest) throws TimeStampTokenGenerationException {
```

Comment:
Maybe keep the more generic name `getResponse` since the base class shouldn't be tied to HTTP